altamisa/isatab/write_assay_study.py

```diff
@@ -0,0 +1,355 @@
+# -*- coding: utf-8 -*-
+"""This module contains code for the parsing of assay and study files.
+"""
+
+
+from __future__ import generator_stop
+import csv
+import os
+from typing import NamedTuple, TextIO
+
+from ..constants.table_headers import *  # noqa: F403
+from ..constants.table_tokens import TOKEN_UNKNOWN
+from ..exceptions import WriteIsatabException
+from .headers import AssayHeaderParser, StudyHeaderParser
+from .models import Material, Process
+
+
+__author__ = (
+    "Manuel Holtgrewe <manuel.holtgrewe@bihealth.de>, "
+    "Mathias Kuhring <mathias.kuhring@bihealth.de>"
+)
+
+
+# Graph traversal -------------------------------------------------------
+
+
+class Digraph:
+    """Simple class encapsulating directed graph with vertices and arcs"""
+
+    def __init__(self, vertices, arcs, starting_type: str):
+        self.vertices = vertices
+        self.arcs = arcs
+        self.v_by_name = {v.unique_name: v for v in self.vertices}
+        self.a_by_name = {(a[0], a[1]) for a in self.arcs}
+        self.source_names = [
+            v.unique_name for v in self.vertices if (hasattr(v, "type") and v.type == starting_type)
+        ]
+        self.outgoing = {}
+
+        for s_name, t_name in self.a_by_name:
+            self.outgoing.setdefault(s_name, []).append(t_name)
+
+
+class UnionFind:
+    """Union-Find (disjoint set) data structure allowing to address by vertex
+    name"""
+
+    def __init__(self, vertex_names):
```

Comment:
Please document fields by adding `#:` comments before `self....` asignments

Reply:
same in all other calses

Reply:
Can you please provide me the documentation for the following variables in Class UnionFind:
```
    self._id = list(range(len(vertex_names)))
    self._sz = [1] * len(vertex_names)
```
I didn't took the time yet (and it would probably take me some) to understand what your algorithm is exactly doing in detail, so I can't currently comment on those two variables.

Reply:
```python
    #: pointer to the containing sets
    self._id = list(range(len(vertex_names)))
    #: size of the set (_sz[_id[v]] is the size of the set that contains v)
    self._sz = [1] * len(vertex_names)
```

Reply:
Actually, it's just Union-Find :)

https://en.wikipedia.org/wiki/Disjoint-set_data_structure

Really one of the pearls of algorithmics!

Reply:
Thanks. Might have covered that in bachelors ~10 years ago, will certainly have look.